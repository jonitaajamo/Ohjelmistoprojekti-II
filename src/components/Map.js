import React, { Component } from "react";
import { geoMercator, geoPath } from "d3-geo";
import { feature } from "topojson-client";
import Tooltip from "./Tooltip";

class Map extends Component {
  constructor() {
    super();
    this.state = {
      worldData: [],
      countryHover: false,
      activeCountry: "",
      clicked: false
    };
  }

  toggleHover(i) {
    this.setState({
      countryHover: !this.state.countryHover,
      activeCountry: this.state.worldData[i].id
    });
  }

  componentDidMount() {
    fetch("https://unpkg.com/world-atlas@1.1.4/world/110m.json")
      .then(response => response.json())
      .then(worldData =>
        this.setState({
          worldData: feature(worldData, worldData.objects.countries).features
        })
      )
      .catch(err => console.error(err));
  }

  onClick() {
    this.setState({ clicked: true });
  }

  renderTooltip() {
    return this.state.clicked ? <Tooltip country={this.state.activeCountry}/> : null;
  }

  render() {
    let countryStyle = {
      fill: "#CCCCCC",
      stroke: "#000000",
      strokeWidth: "0.5px"
    };

    let activeStyle = {
      fill: "pink",
      stroke: "#000000",
      strokeWidth: "0.5px"
    };

    let svgStyle = {
    }

    let containerStyle = {
      width: "100%",
      display: "block",
      margin: "auto",
      textAlign: "center"
    }

    const projection = geoMercator().scale(100);
    const pathGenerator = geoPath().projection(projection);
    const countries = this.state.worldData.map((d, i) => (
      <path
        style={
          this.state.countryHover &&
          this.state.activeCountry === this.state.worldData[i].id
            ? activeStyle
            : countryStyle
        }
        key={"path" + i}
        d={pathGenerator(d)}
        className="countries"
        onClick={() => this.onClick()}
        onMouseOver={() => this.toggleHover(i)}
        onMouseLeave={() => this.toggleHover(i)}
      />
    ));
    return (
      <div style={containerStyle}>
        <svg style={svgStyle} preserveAspectRatio="xMidYMin" viewBox="165 20 800 450">
          {countries}
        </svg>
        <div>{this.renderTooltip()}</div>
      </div>
    );
  }
}

export default Map;
