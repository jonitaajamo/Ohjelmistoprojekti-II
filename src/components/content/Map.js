import React, { Component } from "react";
import { feature } from "topojson-client";
import Country from "./Country";
import Loading from "./Loading";
import { geoPath } from "d3-geo";
import { geoTimes } from "d3-geo-projection";
import {
  ComposableMap,
  ZoomableGroup,
  Geographies,
  Geography
} from "react-simple-maps";

class Map extends Component {
  constructor() {
    super();
    this.state = {
      worldData: [],
      countryNames: [],
      countryHover: false,
      hoveredCountry: "",
      clicked: false,
      activeCountry: "",
      zoom: 1,
      center: [0, 20],
      disableOptimization: false
    };
  }

  componentDidMount() {
    fetch("https://unpkg.com/world-atlas@1.1.4/world/110m.json")
      .then(response => response.json())
      .then(worldData =>
        this.setState({
          worldData: feature(
            worldData,
            worldData.objects.countries
          ).features.sort((a, b) => {
            return a.id - b.id;
          })
        })
      )
      .catch(err => console.error(err));

    fetch(
      "https://raw.githubusercontent.com/tarmeli/Ohjelmistoprojekti-II/master/src/data/countryNames.json"
    )
      .then(response => response.json())
      .then(names =>
        this.setState({
          countryNames: names.countries.sort((a, b) => {
            return a.id - b.id;
          })
        })
      )
      .catch(err => console.error(err));
  }

  toggleHover(i) {
    this.setState({
      countryHover: !this.state.countryHover,
      hoveredCountry: this.state.countryNames[i].name
    });
  }

  projection() {
    return geoTimes()
      .translate([800 / 2, 450 / 2])
      .scale(160);
  }

  onClick(geography, i) {
    const path = geoPath().projection(this.projection());
    const centroid = this.projection().invert(path.centroid(geography));
    this.setState(
      {
        clicked: true,
        activeCountry: this.state.countryNames[i].name,
        zoom: 3,
        center: centroid,
        disableOptimization: true
      },
      () => {
        this.setState({
          disableOptimization: false
        });
      }
    );

    if (this.state.activeCountry === this.state.countryNames[i].name) {
      this.setState({
        clicked: false,
        activeCountry: "",
        zoom: 1,
        center: [0, 20]
      });
    }
  }

  checkCountry() {
    let country;
    if (this.state.clicked) {
      country = this.state.activeCountry;
    } else if (this.state.countryHover) {
      country = this.state.hoveredCountry;
    } else {
      country = "Select a country";
    }
    return country;
  }

  zoomOut() {
    this.setState(
      {
        clicked: false,
        zoom: 1,
        center: [0, 20],
        activeCountry: "",
        disableOptimization: true
      },
      () => {
        this.setState({
          disableOptimization: false
        });
      }
    );
  }

  renderMap() {
    const buttonStyle = {
      margin: "10px",
    };

    const mapGeographies = (
      <Geographies
        disableOptimization={this.state.disableOptimization}
        geography={this.state.worldData}
      >
        {(geographies, projection) =>
          geographies.map((geography, i) => (
            <Geography
              style={{
                default: {
                  fill:
                    this.state.clicked &&
                    this.state.activeCountry === this.state.countryNames[i].name
                      ? "steelblue"
                      : this.state.countryNames[i].data === 10
                        ? "#fcfcfc"
                        : `rgba(200,50,56, ${1 *
                            this.state.countryNames[i].data /
                            10000})`,
                  stroke: "black",
                  strokeWidth: "0.5px",
                  outline: "none"
                },
                hover: {
                  fill: "rgba(200, 50, 56, 0.5)",
                  stroke: "black",
                  strokeWidth: "0.5px",
                  outline: "none"
                },
                pressed: {
                  fill: "tomato",
                  stroke: "black",
                  strokeWidth: "0.5px",
                  outline: "none"
                }
              }}
              key={geography + i}
              geography={geography}
              projection={projection}
              className="countries"
              onClick={() => this.onClick(geography, i)}
              onMouseOver={() => this.toggleHover(i)}
              onMouseLeave={() => this.toggleHover(i)}
            />
          ))
        }
      </Geographies>
    );

    return (
      <article className="tile is-child notification is-paddingless">
        <ComposableMap
          width={800}
          height={450}
          style={{
            width: "100%",
            height: "auto",
            marginBottom: "-62px"
          }}
        >
          <ZoomableGroup center={this.state.center} zoom={this.state.zoom}>
            {mapGeographies}
          </ZoomableGroup>
        </ComposableMap>
        <button
          className="button"
          style={buttonStyle}
          onClick={() => this.zoomOut()}
        >
          Zoom out
        </button>
        <Country country={this.checkCountry()} />
      </article>
    );
  }

  renderLoading() {
    return (
      <article className="tile is-child notification is-paddingless">
        <Loading color="tomato" />
      </article>
    );
  }

  render() {
    if (!this.state.countryNames.length) {
      return this.renderLoading();
    } else {
      return this.renderMap();
    }
  }
}

export default Map;
