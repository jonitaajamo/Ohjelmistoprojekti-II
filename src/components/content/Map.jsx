import React, { Component } from "react";
import Country from "./map/Country";
import Loading from "./map/Loading";
import Timeline from "./map/Timeline";
import Assets from "./map/Assets";
import ButtonGroup from "./map/ButtonGroup";
import { geoPath } from "d3-geo";
import { geoTimes } from "d3-geo-projection";
import {
  ComposableMap,
  ZoomableGroup,
  Geographies,
  Geography
} from "react-simple-maps";

class Map extends Component {
  constructor(props) {
    super(props);
    this.state = {
      isGeographyHovered: false,
      hoveredGeographyName: "",
      isGeographyClicked: true,
      clickedGeographyName: "",
      mapZoomValue: 1,
      mapCenter: [0, 20],
      disableOptimization: false
    };
  }

  toggleHover(i) {
    this.setState({
      isGeographyHovered: !this.state.isGeographyHovered,
      hoveredGeographyName: this.props.geographyNames[i].name
    });
  }

  mapProjection() {
    return geoTimes()
      .translate([800 / 2, 450 / 2])
      .scale(160);
  }

  onGeographyClick(geography, i) {
    const path = geoPath().projection(this.mapProjection());
    const centroid = this.mapProjection().invert(path.centroid(geography));

    this.setState(
      {
        isGeographyClicked: true,
        clickedGeographyName: this.props.geographyNames[i].name,
        mapZoomValue: 3,
        mapCenter: centroid,
        disableOptimization: true
      },
      () => {
        this.setState({
          disableOptimization: false
        });
      }
    );

    if (
      this.state.clickedGeographyName === this.props.geographyNames[i].name &&
      this.state.isGeographyClicked
    ) {
      this.setState({
        isGeographyClicked: false,
        clickedGeographyName: "",
        mapZoomValue: 1,
        mapCenter: [0, 20]
      });
    }

    this.props.fetchGeographyIdFromMap(
      geography.id,
      this.state.isGeographyClicked
    );
  }

  checkGeographyName() {
    let geography;
    if (this.state.isGeographyClicked) {
      geography = this.state.clickedGeographyName;
    } else if (this.state.isGeographyHovered) {
      geography = this.state.hoveredGeographyName;
    } else {
      geography = "Select a country";
    }
    return geography;
  }

  setWeightValuesForHeatmap() {
    const weightDataForMap = [];
    const month = this.props.month;
    const asset =
      this.props.selectedAsset === undefined ? 0 : this.props.selectedAsset;
    const assetClasses = this.props.geographicalWeightData.length
      ? this.props.geographicalWeightData[month].assetClasses[asset].weights
      : [];
    for (let i = 0; i < this.props.geographyNames.length; i++) {
      for (let j = 0; j < assetClasses.length; j++) {
        let id = assetClasses[j].countryId;
        let weight = assetClasses[j].weight;
        if (this.props.geographyNames[i].id === JSON.stringify(id)) {
          weightDataForMap.push(weight);
        }
      }
      if (!weightDataForMap[i]) {
        weightDataForMap.push(0);
      }
    }
    return weightDataForMap;
  }

  zoomHandler(e) {
    switch (e.target.value) {
      case "reset":
        this.setState(
          {
            isGeographyClicked: false,
            mapZoomValue: 1,
            mapCenter: [0, 20],
            clickedGeographyName: "",
            disableOptimization: true
          },
          () => {
            this.setState({
              disableOptimization: false
            });
          }
        );
        break;
      case "zoom-out":
        this.setState(
          {
            mapZoomValue:
              this.state.mapZoomValue === 1
                ? this.state.mapZoomValue
                : this.state.mapZoomValue - 1,
            disableOptimization: true
          },
          () => {
            this.setState({
              disableOptimization: false
            });
          }
        );
        break;
      case "zoom-in":
        this.setState(
          {
            mapZoomValue: this.state.mapZoomValue + 1,
            disableOptimization: true
          },
          () => {
            this.setState({
              disableOptimization: false
            });
          }
        );
        break;
      default:
        break;
    }
  }

  renderMap() {
    const weightData = this.setWeightValuesForHeatmap();
    const optimization =
      this.state.disableOptimization || this.props.disableOptimization;
    const mapGeographies = (
      <Geographies
        disableOptimization={optimization}
        geography={this.props.geographyBorders}
      >
        {(geographies, projection) =>
          geographies.map((geography, i) => (
            <Geography
              style={{
                default: {
                  fill:
                    weightData[i] === 0
                      ? "#fcfcfc"
                      : `rgba(200,50,56, ${weightData[i] * 2})`,
                  stroke:
                    this.state.isGeographyClicked &&
                    this.state.clickedGeographyName ===
                      this.props.geographyNames[i].name
                      ? "rgb(50, 115, 220)"
                      : "black",
                  strokeWidth:
                    this.state.isGeographyClicked &&
                    this.state.clickedGeographyName ===
                      this.props.geographyNames[i].name
                      ? "1.4px"
                      : "0.5px",
                  outline: "none"
                },
                hover: {
                  fill: "rgba(200, 50, 56, 0.5)",
                  stroke:
                    this.state.isGeographyClicked &&
                    this.state.clickedGeographyName ===
                      this.props.geographyNames[i].name
                      ? "rgb(50, 115, 220)"
                      : "black",
                  strokeWidth:
                    this.state.isGeographyClicked &&
                    this.state.clickedGeographyName ===
                      this.props.geographyNames[i].name
                      ? "1.4px"
                      : "0.5px",
                  outline: "none"
                },
                pressed: {
                  fill: "rgb(50, 115, 220)",
                  stroke:
                    this.state.isGeographyClicked &&
                    this.state.clickedGeographyName ===
                      this.props.geographyNames[i].name
                      ? "rgb(50, 115, 220)"
                      : "black",
                  strokeWidth:
                    this.state.isGeographyClicked &&
                    this.state.clickedGeographyName ===
                      this.props.geographyNames[i].name
                      ? "1.4px"
                      : "0.5px",
                  outline: "none"
                }
              }}
              key={geography + i}
              geography={geography}
              projection={projection}
              className="countries"
              onClick={() => this.onGeographyClick(geography, i)}
              onMouseOver={() => this.toggleHover(i)}
              onMouseLeave={() => this.toggleHover(i)}
            />
          ))
        }
      </Geographies>
    );

    return (
      <div className="tile is-ancestor">
        <div className="tile is-vertical is-8">
          <div className="tile is-parent">
            <article className="box tile has-accent is-child notification is-paddingless">
              <div>
                <ComposableMap
                  width={800}
                  height={450}
                  style={{
                    width: "100%",
                    height: "auto"
                  }}
                >
                  <ZoomableGroup
                    center={this.state.mapCenter}
                    zoom={this.state.mapZoomValue}
                    disablePanning="true"
                  >
                    {mapGeographies}
                  </ZoomableGroup>
                </ComposableMap>
                <Country country={this.checkGeographyName()} />
                <ButtonGroup
                  isGeographyClicked={this.state.isGeographyClicked}
                  clickedGeographyName={this.state.clickedGeographyName}
                  mapZoomValue={this.state.mapZoomValue}
                  mapCenter={this.state.mapCenter}
                  disableOptimization={this.state.disableOptimization}
                  zoomHandler={this.zoomHandler.bind(this)}
                />
              </div>
              <Timeline
                length={this.props.length}
                geographicalWeightData={this.props.geographicalWeightData}
                onChange={this.props.onChange}
                onAssetChange={this.props.onAssetChange}
                month={this.props.month}
                onMonthButtonClick={this.props.onMonthButtonClick}
                onAfterChange={this.props.onAfterChange}
              />
            </article>
          </div>
        </div>
        <div className="tile">
          <div className="tile is-parent is-vertical">
            <Assets
              geographyId={this.props.geographyId}
              geographicalWeightData={this.props.geographicalWeightData}
              month={this.props.month}
              isGeographyClicked={this.props.isGeographyClicked}
              clickedGeographyName={this.state.clickedGeographyName}
              currency={this.props.currency}
              topInstruments={this.props.topInstruments}
              selectedAsset={this.props.selectedAsset}
            />
          </div>
        </div>
      </div>
    );
  }

  renderLoading() {
    return (
      <article className="is-child notification is-paddingless">
        <Loading item="map" />
      </article>
    );
  }

  render() {
    if (!this.props.isLoaded) {
      return this.renderLoading();
    } else {
      return this.renderMap();
    }
  }
}

export default Map;
