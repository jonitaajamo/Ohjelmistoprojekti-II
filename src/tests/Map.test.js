import React from "react";
import ReactDOM from "react-dom";
import Map from "../components/content/Map";
import { shallow, mount } from "enzyme";

describe("Map", () => {
  it("renders without crashing", () => {
    const wrapper = shallow(<Map />);
    expect(wrapper).toMatchSnapshot();
  });

  it("initially renders Loading component", () => {
    const wrapper = shallow(<Map />);
    expect(wrapper.find("Loading").length).toEqual(1);
  });

  it("renders map when data is loaded", () => {
    const wrapper = shallow(<Map />);
    wrapper.setState({ geographyBorders: [1, 2] });
    expect(wrapper.find("ComposableMap").length).toEqual(1);
  });

  it("fires event handler when geography is clicked", () => {
    const wrapper = mount(<Map />);
    wrapper.setState({ geographyBorders: [1, 2] });
    wrapper.instance().onGeographyClick = jest.fn();
    wrapper.update();
    wrapper.find("Geography").at(0).simulate("click");
    expect(wrapper.instance().onGeographyClick).toHaveBeenCalled();
  });

  it("zooms out when button is clicked", () => {
    const wrapper = shallow(<Map />);
    wrapper.setState({ geographyBorders: [1, 2], mapZoomValue: 3 });
    wrapper.find("button").simulate("click");
    expect(wrapper.state().mapZoomValue).toBe(1);
  });
});
