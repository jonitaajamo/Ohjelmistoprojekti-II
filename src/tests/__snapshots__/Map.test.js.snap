// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Map renders without crashing 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Map
    isLoaded={false}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <Loading
        item="map"
/>,
      "className": "is-child notification is-paddingless",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "item": "map",
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": "article",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <Loading
          item="map"
/>,
        "className": "is-child notification is-paddingless",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "item": "map",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": "article",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
